File: Firmware/vision/tests/test_track_manager.py
"""§82 — TrackManager unit tests, run where the TrackManager actually lives.

The eight required cases, named after §82's own list: single target; crossing
targets; temporary occlusion; detector dropout; two same-class targets; abrupt
bbox-size change; track creation/retirement; UUID stability.

These were first written in C++, against a C++ TrackManager, because §82 sits in a
chapter about controld. That was the wrong process: §9's TrackSet already carries
`state`, `age_frames`, `display_index` and the counters, so tracks are formed before
publication — and §58's "once per detector frame" is a camera-rate statement, which says
the same thing differently. Two implementations of §10's identity rules would disagree
first time two targets crossed, so there is one, here, and controld reads what it says.

No camera, no motor driver, no hardware (§82's closing line).
"""
import unittest

from vision.protocol import BBoxNorm, Track, TrackSet
from vision.track_manager import TrackManager, TrackManagerConfig, TrackState

FRAME_NS = 33_333_333  # 30 Hz camera


def det(class_id=1, name="person", ax=0.5, ay=0.5, w=0.10, h=0.20, conf=0.9):
    return Track(
        class_id=class_id,
        class_name=name,
        detector_confidence=conf,
        track_confidence=conf,
        bbox=BBoxNorm(ax - w / 2, ay - h / 2, ax + w / 2, ay + h / 2),
        anchor_x=ax,
        anchor_y=ay,
    )


class Harness:
    """Frame counter + clock, so the state ladder can be walked deterministically."""

    def __init__(self, manager=None):
        self.m = manager or TrackManager()
        self.seq = 0
        self.now = 0

    def feed(self, *dets):
        self.seq += 1
        self.now += FRAME_NS
        s = TrackSet(
            frame_sequence=self.seq,
            sensor_timestamp_ns=self.seq * FRAME_NS,
            publish_timestamp_ns=self.seq * FRAME_NS + 8_000_000,  # 8 ms later
            width=1280,
            height=720,
            tracks=list(dets),
        )
        return self.m.update(s, self.now)

    def feed_empty(self, n):
        for _ in range(n):
            self.feed()

    def only(self):
        ts = self.m.tracks()
        assert len(ts) == 1, f"expected exactly one track, got {len(ts)}"
        return ts[0]


class SingleTarget(unittest.TestCase):
    def test_becomes_selectable_after_consistent_frames(self):
        h = Harness()
        h.feed(det())
        uuid = h.only().track_uuid
        self.assertFalse(h.m.is_selectable(uuid),
                         "a first-frame detection is not something to point a "
                         "turret at (§8)")
        for _ in range(3):
            h.feed(det())
        self.assertTrue(h.m.is_selectable(uuid),
                        "consistent frames must confirm it (§8)")
        self.assertTrue(h.m.is_trackable(uuid))
        self.assertEqual(h.m.track_count(), 1, "one target must not become three")


class CrossingTargets(unittest.TestCase):
    def test_identities_survive_the_crossing(self):
        # The failure this pins: one frame of overlap makes the association swap, and
        # from then on the turret chases the wrong human being while the overlay labels
        # stay exactly as they were.
        h = Harness()
        a, b = det(ax=0.30), det(ax=0.40)
        for _ in range(4):
            a.anchor_x -= 0.005
            b.anchor_x += 0.005
            h.feed(a, b)
        tracks = h.m.tracks()
        self.assertEqual(len(tracks), 2)
        left = min(tracks, key=lambda t: t.anchor_x).track_uuid

        for _ in range(20):
            a.anchor_x -= 0.005
            b.anchor_x += 0.005
            h.feed(a, b)
        tracks = h.m.tracks()
        self.assertEqual(len(tracks), 2, "a crossing must not delete a track")
        now_left = min(tracks, key=lambda t: t.anchor_x).track_uuid
        self.assertEqual(now_left, left,
                         "identities swapped when the two crossed — the turret would "
                         "follow the wrong person while the screen said the same thing")


class TemporaryOcclusion(unittest.TestCase):
    def test_survives_and_stays_trackable_but_not_selectable(self):
        h = Harness()
        for _ in range(5):
            h.feed(det())
        uuid = h.only().track_uuid
        self.assertTrue(h.m.is_selectable(uuid))

        h.feed_empty(5)  # a pillar passes between camera and target
        self.assertIsNotNone(h.m.find(uuid), "an occluded track must still exist")
        self.assertEqual(h.m.find(uuid).state, TrackState.OCCLUDED)
        self.assertTrue(h.m.is_trackable(uuid),
                        "§20's coasting needs the selected target to still be a "
                        "subject")
        self.assertFalse(h.m.is_selectable(uuid),
                         "an operator must not be able to newly select something the "
                         "detector is not confirming right now")

        h.feed(det())
        self.assertEqual(h.m.find(uuid).state, TrackState.CONFIRMED)
        self.assertTrue(h.m.is_selectable(uuid))


class DetectorDropout(unittest.TestCase):
    def test_walks_the_ladder_then_retires(self):
        h = Harness(TrackManager(TrackManagerConfig(
            lost_frames=6, reacquire_frames=10)))
        for _ in range(5):
            h.feed(det())
        uuid = h.only().track_uuid

        h.feed_empty(6)
        self.assertIsNotNone(h.m.find(uuid))
        self.assertEqual(h.m.find(uuid).state, TrackState.LOST)
        self.assertFalse(h.m.is_trackable(uuid),
                         "a lost track is not something the controller keeps driving "
                         "toward")

        h.feed_empty(11)
        self.assertIsNone(h.m.find(uuid),
                          "§22: past the reacquisition window the track retires")
        self.assertGreater(h.m.stats.retired, 0)


class TwoSameClassTargets(unittest.TestCase):
    def test_get_different_labels(self):
        # §10: "Person #1 / Person #2", never raw UUIDs, and never one label shared by
        # two people.
        h = Harness()
        for _ in range(5):
            h.feed(det(ax=0.3), det(ax=0.7))
        a, b = sorted(h.m.tracks(), key=lambda t: t.display_index)
        self.assertNotEqual(a.track_uuid, b.track_uuid)
        self.assertNotEqual(a.display_index, b.display_index,
                            "two people cannot share one label on the selector")
        self.assertEqual((a.display_index, b.display_index), (1, 2))
        self.assertEqual(a.class_id, b.class_id)

    def test_label_is_reused_only_after_retirement(self):
        # §10 verbatim, and the reason selection carries the UUID: if a label could be
        # handed out while its previous owner still existed, the operator's "#1" would
        # silently mean a different human being.
        h = Harness(TrackManager(TrackManagerConfig(
            lost_frames=4, reacquire_frames=6)))
        for _ in range(5):
            h.feed(det(ax=0.3), det(ax=0.7))
        first = min(h.m.tracks(), key=lambda t: t.display_index)
        survivor = max(h.m.tracks(), key=lambda t: t.display_index)
        self.assertEqual(first.display_index, 1)

        for _ in range(12):
            h.feed(det(ax=0.7))  # only the survivor keeps being detected
        self.assertEqual(h.m.find(survivor.track_uuid).display_index, 2,
                         "the survivor keeps its label")
        self.assertIsNone(h.m.find(first.track_uuid), "the retired one is gone")

        for _ in range(5):
            h.feed(det(ax=0.7), det(ax=0.15, ay=0.8))
        newcomer = [t for t in h.m.tracks() if t.track_uuid != survivor.track_uuid]
        self.assertTrue(any(t.display_index == 1 for t in newcomer),
                        "a freed label goes to the next track, lowest first")


class AbruptBboxSizeChange(unittest.TestCase):
    def test_does_not_kill_the_track(self):
        # §82's quietest requirement, and the reason association is anchored on the
        # anchor point rather than on IoU: a detector that suddenly returns a fifth of
        # the box must not be read as "this object is gone and a new one appeared
        # here".
        h = Harness()
        for _ in range(5):
            h.feed(det(w=0.30, h=0.60))
        uuid = h.only().track_uuid
        self.assertTrue(h.m.is_selectable(uuid))

        for _ in range(3):
            h.feed(det(w=0.06, h=0.10))

        self.assertIsNotNone(h.m.find(uuid),
                             "the box shrank to a fifth and the identity was thrown "
                             "away")
        self.assertEqual(h.m.find(uuid).state, TrackState.CONFIRMED)
        self.assertEqual(h.m.track_count(), 1,
                         "and a second track was invented in its place")


class TrackCreationAndRetirement(unittest.TestCase):
    def test_counts_move_the_way_the_overlay_needs(self):
        h = Harness(TrackManager(TrackManagerConfig(
            confirm_frames=2, lost_frames=4, reacquire_frames=4)))
        self.assertEqual(h.m.stats.created, 0)
        h.feed(det())
        self.assertEqual(h.m.stats.created, 1)
        for _ in range(2):
            h.feed(det())
        self.assertEqual(h.m.find(h.only().track_uuid).state, TrackState.CONFIRMED)
        h.feed_empty(9)
        self.assertEqual(h.m.track_count(), 0, "retired tracks leave the table")
        self.assertEqual(h.m.stats.retired, 1)

    def test_capacity_never_sacrifices_a_confirmed_track(self):
        # At the cap something has to go. Dropping a confirmed track an operator may
        # have selected, to make room for a detection nobody asked for, would look like
        # the turret misbehaving with nothing in the log to explain it.
        cfg = TrackManagerConfig(max_tracks=4)
        h = Harness(TrackManager(cfg))
        dets = [det(ax=0.1 + 0.1 * i) for i in range(4)]
        for _ in range(5):
            h.feed(*dets)
        self.assertEqual(h.m.track_count(), 4)
        oldest = h.m.tracks()[0].track_uuid

        for _ in range(6):
            h.feed(*dets, det(ax=0.95, ay=0.95))
        self.assertLessEqual(h.m.track_count(), cfg.max_tracks,
                             "§60's cap is a contract, not a hint")
        self.assertIsNotNone(h.m.find(oldest),
                             "a confirmed track was dropped to make room for an "
                             "unproven detection")


class UuidStability(unittest.TestCase):
    def test_identity_survives_every_transition(self):
        # §10. The selection command carries the UUID, so an identity that changes
        # under the operator's finger either drops tracking or — worse, because it is
        # silent — retargets onto somebody else.
        h = Harness(TrackManager(TrackManagerConfig(
            lost_frames=6, reacquire_frames=20)))
        for _ in range(5):
            h.feed(det())
        t = h.only()
        uuid, label = t.track_uuid, t.display_index

        h.feed_empty(6)          # through OCCLUDED and toward LOST
        self.assertIsNotNone(h.m.find(uuid))
        h.feed(det(ax=0.52))     # back inside the reacquisition window
        slot = h.m.find(uuid)
        self.assertIsNotNone(
            slot, "reacquisition must reuse the identity, not mint a new one")
        self.assertEqual(slot.display_index, label)
        self.assertGreater(h.m.stats.reacquired, 0)


class LatencyStamps(unittest.TestCase):
    def test_unknown_is_not_zero_and_the_two_intervals_are_separate(self):
        # §61. "Never received" and "0 ms old" are different claims, and a dashboard
        # that shows the second when the first is true makes a dead detector look like
        # a healthy one with a slightly old frame.
        m = TrackManager()
        self.assertLess(m.stale_ms(1_000_000_000), 0.0)
        self.assertLess(m.sensor_age_ms(1_000_000_000), 0.0)
        self.assertLess(m.publish_to_receive_ms(), 0.0)

        capture = 300 * FRAME_NS
        now = capture + 12_000_000
        s = TrackSet(frame_sequence=300, sensor_timestamp_ns=capture,
                     publish_timestamp_ns=capture + 8_000_000,
                     width=1280, height=720, tracks=[det()])
        m.update(s, now)
        self.assertAlmostEqual(m.stale_ms(now), 0.0)
        self.assertAlmostEqual(m.sensor_age_ms(now), 12.0, places=3)
        self.assertAlmostEqual(
            m.publish_to_receive_ms(), 4.0, places=3,
            msg="the transport delay must be visible on its own, or a slow publisher "
                "and a slow camera are indistinguishable — and they need different "
                "fixes")


if __name__ == "__main__":
    unittest.main()

File: Firmware/control/tests/test_track_manager.cpp
// OpenAutoTurret v3 §82 — TrackManager unit tests.
//
// The eight required cases, named after the spec's own list so a reader can check
// §82 off line by line:
//   single target; crossing targets; temporary occlusion; detector dropout;
//   two same-class targets; abrupt bbox-size change; track creation/retirement;
//   UUID stability.
//
// All synthetic, no camera (§82: "Synthetic tests should not depend on camera
// hardware"). The frame helpers below build TrackSets directly; the association
// rules being pinned here are the ones the overlay and the selection manager are
// built on, so a change to them should make this file argue.
#include "tracks/track_manager.hpp"

#include <gtest/gtest.h>

#include <cstring>

#include "tracks/track_set.hpp"

using namespace ota::tracks;

namespace {

constexpr int64_t kFrameNs = 33'333'333;  // 30 Hz camera

Track make_det(uint16_t class_id, const char* name, float ax, float ay,
               float w = 0.10f, float h = 0.20f, float conf = 0.9f) {
  Track t;
  t.class_id = class_id;
  std::strncpy(t.class_name, name, kClassNameLen - 1);
  t.anchor_x = ax;
  t.anchor_y = ay;
  t.bbox.x_min = ax - w / 2.0f;
  t.bbox.x_max = ax + w / 2.0f;
  t.bbox.y_min = ay - h / 2.0f;
  t.bbox.y_max = ay + h / 2.0f;
  t.detector_confidence = conf;
  t.track_confidence = conf;
  return t;
}

// Feed one frame. `seq` advances the capture clock so the predictor sees a real dt.
void feed(TrackManager& m, int64_t seq, std::initializer_list<Track> dets,
          int64_t now_ns) {
  TrackSet set;
  set.frame_sequence = seq;
  set.sensor_timestamp_ns = seq * kFrameNs;
  set.publish_timestamp_ns = set.sensor_timestamp_ns + 8'000'000;  // 8 ms later
  set.width = 1280;
  set.height = 720;
  for (const Track& d : dets) set.add(d);
  m.update(set, now_ns);
}

// Hold the manager at `now_ns` while advancing frames, so state ladders that are
// measured in frames can be walked without a wall-clock dependency.
void feed_empty(TrackManager& m, int64_t& seq, int n, int64_t& now_ns) {
  for (int i = 0; i < n; ++i) {
    feed(m, ++seq, {}, now_ns += kFrameNs);
  }
}

}  // namespace

TEST(TrackManager, SingleTargetBecomesSelectable) {
  TrackManager m;
  int64_t seq = 0, now = 0;

  feed(m, ++seq, {make_det(0, "person", 0.5f, 0.5f)}, now += kFrameNs);
  const TrackUuid uuid = [&] {
    Track t{};
    m.copy_tracks(&t, 1);
    return t.uuid;
  }();
  ASSERT_TRUE(uuid.valid());
  EXPECT_FALSE(m.is_selectable(uuid))
      << "a first-frame detection is not something to point a turret at (§8)";

  for (int i = 0; i < 3; ++i)
    feed(m, ++seq, {make_det(0, "person", 0.5f, 0.5f)}, now += kFrameNs);
  EXPECT_TRUE(m.is_selectable(uuid))
      << "consistent frames must confirm it (§8 TENTATIVE -> CONFIRMED)";
  EXPECT_TRUE(m.is_trackable(uuid));
  EXPECT_EQ(m.track_count(), 1) << "one target must not become three";
}

TEST(TrackManager, CrossingTargetsKeepTheirIdentities) {
  // Two targets on a collision course, passing within a few pixels of each other.
  // The failure mode this pins: one frame of overlap makes the association swap, and
  // from then on the turret is chasing the wrong human being while the overlay labels
  // stay exactly as they were. Constant-velocity prediction plus mutual-best matching
  // is what keeps each track on its own momentum.
  TrackManager m;
  int64_t seq = 0, now = 0;

  Track a = make_det(0, "person", 0.30f, 0.5f);
  Track b = make_det(0, "person", 0.40f, 0.5f);
  for (int i = 0; i < 4; ++i) {  // establish both, moving toward each other
    a.anchor_x -= 0.005f;
    b.anchor_x += 0.005f;
    feed(m, ++seq, {a, b}, now += kFrameNs);
  }
  Track tracks[2];
  ASSERT_EQ(m.copy_tracks(tracks, 2), 2);
  const TrackUuid left = tracks[0].anchor_x < tracks[1].anchor_x ? tracks[0].uuid
                                                                : tracks[1].uuid;
  const TrackUuid right = left == tracks[0].uuid ? tracks[1].uuid : tracks[0].uuid;

  // Cross them: they meet, overlap, and pass through each other.
  for (int i = 0; i < 20; ++i) {
    a.anchor_x -= 0.005f;
    b.anchor_x += 0.005f;
    feed(m, ++seq, {a, b}, now += kFrameNs);
  }
  ASSERT_EQ(m.copy_tracks(tracks, 2), 2) << "a crossing must not delete a track";
  const TrackUuid now_left = tracks[0].anchor_x < tracks[1].anchor_x ? tracks[0].uuid
                                                                    : tracks[1].uuid;
  EXPECT_EQ(now_left, left)
      << "identities swapped when the two crossed — the turret would now follow the "
         "wrong person while the screen still says the same thing";
  EXPECT_NE(now_left, right);
}

TEST(TrackManager, TemporaryOcclusionSurvivesAndStaysTrackable) {
  TrackManager m;
  int64_t seq = 0, now = 0;
  for (int i = 0; i < 5; ++i)
    feed(m, ++seq, {make_det(0, "person", 0.5f, 0.5f)}, now += kFrameNs);
  Track t{};
  ASSERT_EQ(m.copy_tracks(&t, 1), 1);
  const TrackUuid uuid = t.uuid;
  ASSERT_TRUE(m.is_selectable(uuid));

  feed_empty(m, seq, 5, now);  // a pillar passes between camera and target
  EXPECT_NE(m.find(uuid), nullptr) << "an occluded track must still exist";
  EXPECT_EQ(m.find(uuid)->state, TrackState::Occluded) << "§8";
  EXPECT_TRUE(m.is_trackable(uuid))
      << "§20's coasting needs the selected target to still be a subject";
  EXPECT_FALSE(m.is_selectable(uuid))
      << "but an operator must not be able to newly select something the detector "
         "is not confirming right now";

  feed(m, ++seq, {make_det(0, "person", 0.5f, 0.5f)}, now += kFrameNs);
  EXPECT_EQ(m.find(uuid)->state, TrackState::Confirmed);
  EXPECT_TRUE(m.is_selectable(uuid));
}

TEST(TrackManager, DetectorDropoutWalksTheStateLadderAndRetires) {
  TrackManager m;
  int64_t seq = 0, now = 0;
  for (int i = 0; i < 5; ++i)
    feed(m, ++seq, {make_det(0, "person", 0.5f, 0.5f)}, now += kFrameNs);
  Track t{};
  ASSERT_EQ(m.copy_tracks(&t, 1), 1);
  const TrackUuid uuid = t.uuid;

  feed_empty(m, seq, m.config().lost_frames, now);
  ASSERT_NE(m.find(uuid), nullptr);
  EXPECT_EQ(m.find(uuid)->state, TrackState::Lost) << "§8";
  EXPECT_FALSE(m.is_trackable(uuid))
      << "a lost track is not something the controller keeps driving toward";

  feed_empty(m, seq, m.config().reacquire_frames + 1, now);
  EXPECT_EQ(m.find(uuid), nullptr)
      << "§22: past the reacquisition window the track retires";
  EXPECT_GT(m.stats().retired, 0u);
}

TEST(TrackManager, TwoSameClassTargetsGetDifferentLabels) {
  // §10: "Person #1 / Person #2", not raw UUIDs, and not one label shared by two
  // people. The reuse rule (below) is the other half of the same section.
  TrackManager m;
  int64_t seq = 0, now = 0;
  for (int i = 0; i < 5; ++i) {
    feed(m, ++seq,
         {make_det(0, "person", 0.3f, 0.5f), make_det(0, "person", 0.7f, 0.5f)},
         now += kFrameNs);
  }
  Track tracks[4];
  ASSERT_EQ(m.copy_tracks(tracks, 4), 2);
  EXPECT_NE(tracks[0].uuid, tracks[1].uuid);
  EXPECT_NE(tracks[0].display_index, tracks[1].display_index)
      << "two people cannot share one label on the selector";
  EXPECT_EQ(tracks[0].display_index, 1);
  EXPECT_EQ(tracks[1].display_index, 2);
  EXPECT_EQ(tracks[0].class_id, tracks[1].class_id);
}

TEST(TrackManager, DisplayIndexIsReusedOnlyAfterRetirement) {
  // §10 verbatim. This is the reason selection must carry the UUID: if an index
  // could be handed out while its previous owner still existed, the operator's
  // "#1" would silently mean a different human being.
  TrackManager m;
  int64_t seq = 0, now = 0;
  for (int i = 0; i < 5; ++i) {
    feed(m, ++seq,
         {make_det(0, "person", 0.3f, 0.5f), make_det(0, "person", 0.7f, 0.5f)},
         now += kFrameNs);
  }
  Track tracks[2];
  ASSERT_EQ(m.copy_tracks(tracks, 2), 2);
  const TrackUuid first = tracks[0].display_index == 1 ? tracks[0].uuid : tracks[1].uuid;

  // Kill the #1 track only: keep feeding a detection at the other anchor.
  TrackUuid survivor;
  survivor = first == tracks[0].uuid ? tracks[1].uuid : tracks[0].uuid;
  const float keep_x = tracks[0].display_index == 1 ? 0.7f : 0.3f;
  for (int i = 0; i < m.config().reacquire_frames + m.config().lost_frames + 2; ++i)
    feed(m, ++seq, {make_det(0, "person", keep_x, 0.5f)}, now += kFrameNs);

  ASSERT_EQ(m.find(survivor)->display_index, 2) << "the survivor keeps its label";
  EXPECT_EQ(m.find(first), nullptr) << "the retired one is gone";

  // A third person appears and takes the freed label — legitimately, because #1's
  // owner is fully retired.
  for (int i = 0; i < 5; ++i) {
    feed(m, ++seq,
         {make_det(0, "person", keep_x, 0.5f), make_det(0, "person", 0.15f, 0.8f)},
         now += kFrameNs);
  }
  Track after[3];
  const int n = m.copy_tracks(after, 3);
  bool newcomer_has_one = false;
  for (int i = 0; i < n; ++i)
    if (after[i].uuid != survivor && after[i].display_index == 1)
      newcomer_has_one = true;
  EXPECT_TRUE(newcomer_has_one)
      << "a freed index should be handed to the next track, lowest first";
}

TEST(TrackManager, AbruptBboxSizeChangeDoesNotKillTheTrack) {
  // §82's quietest requirement and the reason association is anchored on the anchor
  // point rather than on IoU. A detector that suddenly returns half the box (target
  // crouches, occluder clips the legs, model has an off frame) must not be read as
  // "this object is gone and a new one appeared here".
  TrackManager m;
  int64_t seq = 0, now = 0;
  for (int i = 0; i < 5; ++i)
    feed(m, ++seq, {make_det(0, "person", 0.5f, 0.5f, 0.30f, 0.60f)},
         now += kFrameNs);
  Track t{};
  ASSERT_EQ(m.copy_tracks(&t, 1), 1);
  const TrackUuid uuid = t.uuid;
  ASSERT_TRUE(m.is_selectable(uuid));

  for (int i = 0; i < 3; ++i)
    feed(m, ++seq, {make_det(0, "person", 0.5f, 0.5f, 0.06f, 0.10f)},
         now += kFrameNs);

  ASSERT_NE(m.find(uuid), nullptr)
      << "the box shrank to a fifth and the identity was thrown away";
  EXPECT_EQ(m.find(uuid)->state, TrackState::Confirmed);
  EXPECT_EQ(m.track_count(), 1) << "and a second track was invented in its place";
}

TEST(TrackManager, UuidStaysStableAcrossEveryTransition) {
  // §10. The selection command carries the UUID, so an identity that changes under
  // the operator's finger would either drop tracking or — worse, because it is
  // silent — retarget onto somebody else.
  TrackManager m;
  int64_t seq = 0, now = 0;
  for (int i = 0; i < 5; ++i)
    feed(m, ++seq, {make_det(0, "person", 0.5f, 0.5f)}, now += kFrameNs);
  Track t{};
  ASSERT_EQ(m.copy_tracks(&t, 1), 1);
  const TrackUuid uuid = t.uuid;
  const uint16_t label = t.display_index;

  feed_empty(m, seq, 6, now);                       // through OCCLUDED
  ASSERT_NE(m.find(uuid), nullptr);
  feed(m, ++seq, {make_det(0, "person", 0.52f, 0.5f)}, now += kFrameNs);  // back
  ASSERT_NE(m.find(uuid), nullptr);
  EXPECT_EQ(m.find(uuid)->uuid, uuid);
  EXPECT_EQ(m.find(uuid)->display_index, label);

  // A re-identification inside the reacquisition window keeps the same identity too
  // (§21), which is what lets a selection survive being briefly lost.
  feed_empty(m, seq, m.config().lost_frames + 2, now);
  ASSERT_NE(m.find(uuid), nullptr) << "still inside the reacquire window";
  feed(m, ++seq, {make_det(0, "person", 0.52f, 0.5f)}, now += kFrameNs);
  ASSERT_NE(m.find(uuid), nullptr)
      << "reacquisition must reuse the identity, not mint a new one";
  EXPECT_GT(m.stats().reacquired, 0u);
}

TEST(TrackManager, LatencyStampsAreReportedAndUnknownIsNotZero) {
  // §61. "Never received" and "0 ms old" are different claims, and a dashboard that
  // shows the second when the first is true is what makes a dead detector look like
  // a healthy one with an old frame.
  TrackManager m;
  EXPECT_LT(m.stale_ms(1'000'000'000), 0.0);
  EXPECT_LT(m.sensor_age_ms(1'000'000'000), 0.0);
  EXPECT_LT(m.publish_to_receive_ms(), 0.0);

  // Three stamps, two intervals (§61): captured at T, let go by visiond at T+8 ms,
  // received by controld at T+12 ms. The first version of this test fed a receipt
  // timestamp *earlier* than the publish stamp and failed — which was the
  // implementation refusing to publish a negative transport delay, and the test
  // inventing a frame that arrives before it is sent. The guard is the behaviour;
  // the numbers below are the ones a real pipeline could produce.
  const int64_t capture = 300 * kFrameNs;               // sensor_timestamp_ns
  const int64_t now = capture + 12'000'000;             // received 12 ms later
  feed(m, /*seq=*/300, {make_det(0, "person", 0.5f, 0.5f)}, now);
  EXPECT_DOUBLE_EQ(m.stale_ms(now), 0.0) << "this frame just arrived";
  EXPECT_NEAR(m.sensor_age_ms(now), 12.0, 0.001)
      << "how stale the observation is, measured from the camera's own clock";
  EXPECT_NEAR(m.publish_to_receive_ms(), 4.0, 0.001)
      << "the transport delay has to be visible on its own, or a slow publisher and "
         "a slow camera are indistinguishable — and they need different fixes";
}

TEST(TrackManager, CapacityPressureNeverSacrificesASelectedTrack) {
  // At kMaxTracks the manager has to drop something. Dropping a confirmed track the
  // operator may have selected, to make room for a detection nobody asked for, is the
  // one option that would look like the turret misbehaving with nothing in the log.
  TrackManager m;
  int64_t seq = 0, now = 0;
  Track dets[kMaxTracks];
  for (int i = 0; i < kMaxTracks; ++i)
    dets[i] = make_det(0, "person", 0.01f + 0.02f * i, 0.5f);
  for (int round = 0; round < 5; ++round) {
    TrackSet set;
    set.frame_sequence = seq;
    set.sensor_timestamp_ns = seq * kFrameNs;
    for (int i = 0; i < kMaxTracks; ++i) set.add(dets[i]);
    m.update(set, now += kFrameNs);
    ++seq;
  }
  ASSERT_EQ(m.track_count(), kMaxTracks);
  Track first{};
  const int n = m.copy_tracks(&first, 1);
  ASSERT_EQ(n, 1);
  const TrackUuid oldest = first.uuid;

  // One more track on top of a full table.
  for (int round = 0; round < 6; ++round) {
    TrackSet set;
    set.frame_sequence = seq;
    set.sensor_timestamp_ns = seq * kFrameNs;
    for (int i = 0; i < kMaxTracks; ++i) set.add(dets[i]);
    set.add(make_det(0, "person", 0.99f, 0.99f));
    m.update(set, now += kFrameNs);
    ++seq;
  }
  EXPECT_EQ(m.track_count(), kMaxTracks) << "the cap is a contract, not a hint";
  EXPECT_NE(m.find(oldest), nullptr)
      << "a confirmed track was dropped to make room for an unproven detection";
}
